sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -597,14 +738,87 @@ def _transaction(self, transaction_type: TransactionType) -> t.Generator[None, N
             yield
 
 
+def _rows_to_snapshot_intervals(rows: t.Iterable[t.Tuple]) -> t.List[SnapshotIntervals]:
+    intervals: t.Dict[t.Tuple[str, str, str], Intervals] = defaultdict(list)
+    dev_intervals: t.Dict[t.Tuple[str, str, str], Intervals] = defaultdict(list)
+    for row in rows:
+        _, name, identifier, version, start, end, is_dev, is_removed = row
+        intervals_key = (name, identifier, version)
+        target_intervals = intervals if not is_dev else dev_intervals
+        if is_removed:
+            target_intervals[intervals_key] = remove_interval(
+                target_intervals[intervals_key], start, end
+            )
+        else:
+            target_intervals[intervals_key] = merge_intervals(
+                [*target_intervals[intervals_key], (start, end)]
+            )
+
+    result = []
+    for name, identifier, version in {**intervals, **dev_intervals}:
+        result.append(
+            SnapshotIntervals(
+                name=name,
+                identifier=identifier,
+                version=version,
+                intervals=intervals.get((name, identifier, version), []),
+                dev_intervals=dev_intervals.get((name, identifier, version), []),
+            )
+        )
+
+    return result
+
+
+def _intervals_to_df(
+    snapshots: t.Iterable[Snapshot],
+    start: TimeLike,
+    end: TimeLike,
+    is_dev: bool,
+    is_removed: bool,
+) -> pd.DataFrame:
+    return pd.DataFrame(
+        [
+            _interval_to_df(
+                snapshot,
+                *snapshot.inclusive_exclusive(start, end),
+                is_dev=is_dev,
+                is_removed=is_removed,
+            )
+            for snapshot in snapshots
+        ]
+    )
+
+
+def _interval_to_df(
+    snapshot: t.Union[Snapshot, SnapshotIntervals],
+    start_ts: int,
+    end_ts: int,
+    is_dev: bool = False,
+    is_removed: bool = False,
+    is_compacted: bool = False,
+) -> t.Dict[str, t.Any]:
+    return {
+        "id": random_id(),
+        "created_ts": now_timestamp(),
+        "name": snapshot.name,
+        "identifier": snapshot.identifier,
+        "version": snapshot.version,
+        "start_ts": start_ts,
+        "end_ts": end_ts,
+        "is_dev": is_dev,
+        "is_removed": is_removed,
+        "is_compacted": is_compacted,
```

Comment:
what is is_compacted for?

Reply:
Whether the record was produced as a result of compaction.